rllib/algorithms/ppo/tf/ppo_tf_rl_module.py

```diff
@@ -24,194 +27,217 @@ class PPOTfRLModule(TfRLModule):
+        return self.input_specs_exploration()
 
-    @override(TfRLModule)
-    def output_specs_inference(self) -> List[str]:
-        return [SampleBatch.ACTION_DIST]
+    @override(RLModule)
+    def output_specs_inference(self) -> SpecDict:
+        return SpecDict({SampleBatch.ACTION_DIST: Deterministic})
 
-    @override(TfRLModule)
-    def _forward_inference(self, batch) -> Mapping[str, Any]:
+    @override(RLModule)
+    def _forward_inference(self, batch: NestedDict) -> Mapping[str, Any]:
         output = {}
 
-        encoder_out = self.encoder(batch)
-        if STATE_OUT in encoder_out:
-            output[STATE_OUT] = encoder_out[STATE_OUT]
+        # TODO (Artur): Remove this once Policy supports RNN
+        batch[STATE_IN] = tree.map_structure(
+            lambda x: tf.stack([x] * len(batch[SampleBatch.OBS])),
+            self.encoder.get_initial_state(),
+        )
+        batch[SampleBatch.SEQ_LENS] = tf.ones(len(batch[SampleBatch.OBS]))
 
-        action_logits = self.pi(encoder_out[ENCODER_OUT])
+        encoder_outs = self.encoder(batch)
+        # TODO (Artur): Un-uncomment once Policy supports RNN
+        # output[STATE_OUT] = encoder_outs[STATE_OUT]
 
+        # Actions
+        action_logits = self.pi(encoder_outs[ENCODER_OUT][ACTOR])
         if self._is_discrete:
             action = tf.math.argmax(action_logits, axis=-1)
         else:
             action, _ = tf.split(action_logits, num_or_size_splits=2, axis=1)
-
         action_dist = Deterministic(action, model=None)
         output[SampleBatch.ACTION_DIST] = action_dist
 
         return output
 
-    @override(TfRLModule)
-    def input_specs_exploration(self) -> List[str]:
-        return self.input_specs_inference()
+    @override(RLModule)
+    def input_specs_exploration(self):
+        # TODO (Artur): Infer from encoder specs as soon as Policy supports RNN
+        return SpecDict()
 
-    @override(TfRLModule)
-    def output_specs_exploration(self) -> List[str]:
+    @override(RLModule)
+    def output_specs_exploration(self) -> SpecDict:
         return [
             SampleBatch.ACTION_DIST,
             SampleBatch.VF_PREDS,
             SampleBatch.ACTION_DIST_INPUTS,
         ]
 
-    @override(TfRLModule)
+    @override(RLModule)
     def _forward_exploration(self, batch: NestedDict) -> Mapping[str, Any]:
+        """PPO forward pass during exploration.
+
+        Besides the action distribution, this method also returns the parameters of the
+        policy distribution to be used for computing KL divergence between the old
+        policy and the new policy during training.
+        """
         output = {}
-        encoder_out = self.encoder(batch)
-        if STATE_OUT in encoder_out:
-            output[STATE_OUT] = encoder_out[STATE_OUT]
 
-        action_logits = self.pi(encoder_out[ENCODER_OUT])
-        vf = self.vf(encoder_out[ENCODER_OUT])
+        # TODO (Artur): Remove this once Policy supports RNN
```

Comment:
For the first one: This does not work in light of the ActorCriticEncoder since this one already divides inputs between Actor and Critic and therefore already checks for the existence of these keys.

Reply:
(We'll remove this code anyway in https://github.com/ray-project/ray/pull/32723)

Reply:
For the second one: Can we leave this here, since it'll be removed very soon? Let me know if I should do a deepcopy here instead.
[This PR](https://github.com/ray-project/ray/pull/32723) removes it.

Reply:
Are you sure we can get RNN stuff in very soon 🤣 ?? I wouldn't be too optimistic. But let's role with it for now.

Reply:
@kouroshHakha Soon is relative, but in terms of PRs, it will be one of the next ones, right? It's what I'm working on now. 🔨